rlog/2024-12-23-merkle-tree.mdx

```diff
@@ -0,0 +1,213 @@
+Through this iterative construction, we reach the root of the tree.
+The value contained in the root node is called the root hash.
+The root hash is a compressed representation of the digest $D$.
+
+![Image 2](/img/vac101/vac101_merkle_tree.png)
+
+Each node in the Merkle tree is computed by taking a hash.
+Since a binary tree with $2^n$ leaves has $2^{n+1}-1$ nodes,
+then we need to evaluate $2^{n+1}-1$ hashes to construct the Merkle tree.
+
+### Merkle tree intregrity
+
+A large quantity of data can be compressed to a single hash value.
+A natural question to ask is: could a clever party find another digest that yields a Merkle tree with the same root hash?
+If possible, this would compromise the ledger since the blockchain's history could be altered.
+Fortunately, Merkle trees are quite secure.
+In fact, Merkle trees can be used to both bind and hide a digest.
+
+The Merkle tree is able to bind a digest with one of the properties of hash functions (see our previous Vac 101 [[3](https://vac.dev/rlog/vac101-fiat-shamir#hash-functions)] for information on hash functions).
+A hash function is collision resistant; it is infeasible for a malicious party to find two values share the same hash value.
+
+This collision resistance property, essentially, fixes the input to each leaf and into their parent, their parent's parent, and so on.
+
+In certain applications,
+it may be desirable for the digest of a Merkle tree to be kept confidential.
+This is achieved with the preimage resistant propery of hash functions.
+A hash function is preimage resistant provided that it is difficult reverse the hashing operation. 
+It would be necessary for a malicious party to find preimages to each node starting from the root node to determine the original digest.
+
+Now, we see that Merkle trees are secured structures that are tamper resistant.
+
+### Proof of membership
+An interesting and critical property of Merkle trees is their ability to prove that any piece of data is part of its digest.
+This can be done with logarithmic storage and logarithmic computation time.
+
+Suppose that we want to show that data $\ell$ is part of the Merkle tree's digest.
+Additionally, suppose that $\mathsf{hash}$ is the hash function used to construct the tree.
+We assume that the hash function $\mathsf{hash}$ can be computed in constant-time for any input.
+
+Suppose that a prover provides data $\ell$ to a verifier,
+and tells the verifier that $\ell$ corresponds to the $i$th leaf of the Merkle tree.
+For the verifier to be convinced that $\ell$ is part of the digest, he needs to be able to construct the tree's root hash using $\mathsf{hash}$, $i$, $\ell$ and some additional information from the prover.
+Specifically, the prover must provide the sibling hashes for each value that the verifier can compute.
+This enables the verifier to compute the parents of the siblings that the prover provides and the values that he was able to produce himself.
```

Comment:
"and the values that he was able to produce himself" is not clear for me. Perhaps, it is better to remove this and add "The last of the computed parents is the root."

Reply:
Added "The last of the computed parents is the root." as the next sentence as it adds the conclusion to the process.

The "and the values that he was able to produce himself" is part of the hash to compute the parents.

Reply:
Actually,  the values that he was able to produce himself" are also hashes of the "parents of the siblings that the prover provides" (except for the leaf hash), if I understood you correctly.

Reply:
The prover provides the siblings along the path, and the path itself is produced using the entry ell and the siblings with consecutive hashings.

Reply:
Nevertheless, "the path itself", which is computed by the verifier, consists of "parents of the siblings that the prover provides", so "the values that he was able to produce himself" are referring to the things, which have already been mentioned.